sklearn/model_selection/tests/test_validation.py

```diff
@@ -1259,61 +1261,220 @@ def test_cross_val_predict_sparse_prediction():
+    for train, test in cv.split(X, y_enc):
+        est = clone(est).fit(X[train], y_enc[train])
+        fold_preds = getattr(est, method)(X[test])
+        for i_col in range(n_targets):
+            fold_cols = np.unique(y_enc[train][:, i_col])
+            if expected_preds[i_col].ndim == 1:
+                # Decision function with <=2 classes
+                expected_preds[i_col][test] = fold_preds[i_col]
+            else:
+                idx = np.ix_(test, fold_cols)
+                expected_preds[i_col][idx] = fold_preds[i_col]
 
-        # Naive loop (should be same as cross_val_predict):
-        for train, test in kfold.split(X, y):
-            est.fit(X[train], y[train])
-            expected_predictions[test] = func(X[test])
+    # Check actual outputs for several representations of y
+    for tg in [y, y + 1, y - 2, y.astype('str')]:
+        cv_predict_output = cross_val_predict(est, X, tg, method=method, cv=cv)
+        assert_array_equal_maybe_list(cv_predict_output, expected_preds)
 
-        predictions = cross_val_predict(est, X, y, method=method,
-                                        cv=kfold)
-        assert_array_almost_equal(expected_predictions, predictions)
 
-        # Test alternative representations of y
-        predictions_y1 = cross_val_predict(est, X, y + 1, method=method,
-                                           cv=kfold)
-        assert_array_equal(predictions, predictions_y1)
+def assert_array_equal_maybe_list(x, y):
+    # If x and y are lists of arrays, compare arrays individually.
+    if isinstance(x, list):
+        for i in range(len(x)):
+            assert_array_equal(x[i], y[i])
+    else:
+        assert_array_equal(x, y)
 
-        predictions_y2 = cross_val_predict(est, X, y - 2, method=method,
-                                           cv=kfold)
-        assert_array_equal(predictions, predictions_y2)
 
-        predictions_ystr = cross_val_predict(est, X, y.astype('str'),
-                                             method=method, cv=kfold)
-        assert_array_equal(predictions, predictions_ystr)
+def test_cross_val_predict_with_method_binary():
+    # This test includes the decision_function with two classes.
+    # This is a special case: it has only one column of output.
+    X, y = make_classification(n_classes=2, random_state=0)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        est = LogisticRegression()
+        check_cross_val_predict_with_method_binary(est, X, y, method)
 
 
-def test_cross_val_predict_with_method():
-    check_cross_val_predict_with_method(LogisticRegression())
+def test_cross_val_predict_with_method_multiclass():
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    X, y = shuffle(X, y, random_state=0)
+    for method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        est = LogisticRegression()
+        check_cross_val_predict_with_method_multiclass(est, X, y, method)
```

Comment:
perhaps also try this with string `y`?

Reply:
The `check_cross_val_predict_multiclass` called here goes through multiple transformations of `y`, including casting it as a string before running `cross_val_predict`. I think the string case is covered.